src/buf/fixed/handle.rs

```diff
@@ -44,59 +42,48 @@ impl Drop for FixedBuf {
         // maintained accordingly to the safety contracts on
         // Self::new and IoBufMut.
         unsafe {
-            registry.check_in(self.index, self.buf.len());
+            registry.check_in(self.buf.index, self.buf.init_len);
         }
     }
 }
 
 impl FixedBuf {
     // Safety: Validity constraints must apply to CheckedOutBuf members:
-    // - iovec must refer to an array allocated by Vec<u8>;
     // - the array will not be deallocated until the buffer is checked in;
     // - the data in the array must be initialized up to the number of bytes
     //   given in init_len.
-    pub(super) unsafe fn new(registry: Rc<RefCell<dyn FixedBuffers>>, data: CheckedOutBuf) -> Self {
-        let CheckedOutBuf {
-            iovec,
-            init_len,
-            index,
-        } = data;
-        let buf = Vec::from_raw_parts(iovec.iov_base as _, init_len, iovec.iov_len);
-        FixedBuf {
-            registry,
-            buf: ManuallyDrop::new(buf),
-            index,
-        }
+    pub(super) unsafe fn new(registry: Rc<RefCell<dyn FixedBuffers>>, buf: CheckedOutBuf) -> Self {
+        FixedBuf { registry, buf }
     }
 
     /// Index of the underlying registry buffer
     pub fn buf_index(&self) -> u16 {
-        self.index
+        self.buf.index
     }
 }
 
 unsafe impl IoBuf for FixedBuf {
     fn stable_ptr(&self) -> *const u8 {
-        self.buf.as_ptr()
+        self.buf.iovec.iov_base as _
     }
 
     fn bytes_init(&self) -> usize {
-        self.buf.len()
+        self.buf.init_len
     }
 
     fn bytes_total(&self) -> usize {
-        self.buf.capacity()
+        self.buf.iovec.iov_len
     }
 }
 
 unsafe impl IoBufMut for FixedBuf {
     fn stable_mut_ptr(&mut self) -> *mut u8 {
-        self.buf.as_mut_ptr()
+        self.buf.iovec.iov_base as _
     }
 
     unsafe fn set_init(&mut self, pos: usize) {
-        if self.buf.len() < pos {
-            self.buf.set_len(pos)
+        if self.buf.iovec.iov_len >= pos {
+            self.buf.init_len = pos
```

Comment:
I'm still confused by this, meaning it doesn't look right to me. Am I the one looking at this wrong? The original code this PR is replacing was growing the value of the Vec<u8> length by comparing the new pos to the vec length (the vec's length was being used to track the amount initialized). This appears to be using the io_vec's length as the comparison which isn't really the question, the io_vec's length serves as the buffer's capacity I believe. Shouldn't it read
```rust
if self.buf.init_len < pos {
    self.buf.init_len = pos
}
// or
self.buf.init_len = cmp::max(self.buf.init_len, pos);
```

Reply:
The check I've put in is simply a bounds check, i.e, that its valid to set the bytes_init to such the given value. The previous version had no such bounds check. Te previous one _is_ consisnte with the other instances of IoBuf in the crate, of which there are 2:

https://github.com/tokio-rs/tokio-uring/blob/d2f1494d1ab92bd7355a2af6069dd5e4ad5f3415/src/buf/io_buf_mut.rs#L37-L47

https://github.com/tokio-rs/tokio-uring/blob/d2f1494d1ab92bd7355a2af6069dd5e4ad5f3415/src/buf/io_buf_mut.rs#L50-L60

I'm going to argue again that not allowing a buffer to mark fewer bytes to be marked init after a read is just weird.

1. Firstly, we're dealing in `u8`'s. I'm not convinced that there is such as thing as an uninitialized u8. I'll await someone to point out some deeply unintuitive optimization llvm does here...

2. What if I want to use [std::ptr::copy](https://doc.rust-lang.org/std/ptr/fn.copy.html) between buffers? I can't mark the target buffer as less init, although it is possible for bytes from the src to be "un-init"

3. Its really unintuitive that after some read operation `bytes_init` doesn't reflect the value actually read. Certainly, looking at any of the read-like ops, i.e
 
https://github.com/tokio-rs/tokio-uring/blob/d2f1494d1ab92bd7355a2af6069dd5e4ad5f3415/src/io/read.rs#L55-L60

you'd be forgiven for assuming the buffer init value was the number of bytes actually read. 

4. Its not repeatable. If i read hello world from a file into 2 different buffers which may have been previously used, I've no idea what the bytes_init value would be, and they certainly have no guarantees of being equal.


If I'm told I'm just wrong, I'll change this, but I still hold its not a sane API.

Reply:
Actually, I'm irritated enough by this to issue a PR. It can have a discussion in its own right

Reply:
Will change here, discuss there